### PR TITLE
Implement optimistic response for Tempo field updates

### PR DESCRIPTION
**Pull request by qu8n**
This PR implements the optimistic update behavior for the Cohort Sample update workflow.

Before:

https://github.com/mskcc/smile-dashboard/assets/86090707/9e6ea35c-f0ed-4e37-8869-68ed81b2d180

After:

https://github.com/mskcc/smile-dashboard/assets/86090707/a7180b13-6565-4aa2-aeac-384c8b1813ce

### CODE REVIEW

File: frontend/src/components/UpdateModal.tsx [hunk 1]
```diff
@@ -27,23 +27,18 @@ export function UpdateModal({
...
 
   useEffect(() => {
     onOpen && onOpen();
     // eslint-disable-next-line
   }, []);
 
-  useEffect(() => {
-    setRowData(changes);
-  }, [changes]);
-
```
Comment (qu8n) [inline frontend/src/components/UpdateModal.tsx:46]: Cleaned up unneeded uses of `useState`.

File: frontend/src/index.tsx [hunk 2]
```diff
@@ -21,6 +21,9 @@ const cache = new InMemoryCache({
     SampleMetadata: {
       keyFields: ["primaryId"],
     },
+    Tempo: {
+      keyFields: ["smileTempoId"],
+    },
```
Comment (qu8n) [inline frontend/src/index.tsx:26]: Configured GraphQL on the frontend to cache Tempo node data based on their `smileTempoId` (required for optimistic response).

File: graphql-server/src/schemas/neo4j.ts
Comment (qu8n) [inline graphql-server/src/schemas/neo4j.ts:1]: This file contains the majority of new changes in this PR, but most changes are either (1) refactoring/moving code around to reduce code repeats and (2) adding missing types. There is one meaningful change that I'm going to highlight and comment on below.

File: graphql-server/src/schemas/neo4j.ts [hunk 3]
```diff
@@ -51,41 +58,85 @@ export async function buildNeo4jDbSchema() {
...
+
+        Object.keys(changedFields).forEach((key) => {
+          const sample = updatedSamples.data.samples[0][sampleKeyForUpdate] as
+            | SampleMetadata[]
+            | Tempo[];
+          if (Array.isArray(sample) && sample.length > 0) {
+            sample[0][key as keyof typeof sample[0]] =
+              changedFields[key as keyof typeof changedFields];
+          }
+        });
```
Comment (qu8n) [inline graphql-server/src/schemas/neo4j.ts:123]: This code block takes the queried sample data (see `query Samples` in `operations.graphql` for the shape of this data), updates it with changes from users, and then returns it to the frontend. This is one of the requirements for optimistic response and was previously missing for the Cohort Sample update flow.

**Review by ao508 — APPROVED** [review]

**Pull request merged**